### ISSUE TITLE
Technical research spike on ElasticSearch for schedule data

### ISSUE THREAD

**Issue by noahmanger** [18F/openFEC#2373]
This is a task to do a technical research spike on the feasibility of using ES for our itemized schedule A and B data. The goal is to come out of this knowing if this is a direction we want to pursue, and if so, what potential constraints we may need to work within.

When we discussed this in grooming we identified a few key questions for this research:

1. Can cloud.gov handle this volume of data?
2. Can this support nightly indexing to go with the nightly refresh?
3. Can it support all of the filters we currently offer? If not, how many can it support?
4. Can it support downloads?

Someone also raised the possibility of there being a plugin for keeping ES in sync with Postgres, so we should look into that too.

**Comment by ccostino** [comment]
Before we go down the road of Elasticsearch, I want to propose something different, or an alternative path to look at first.  I think it'd be worth our while to investigate a bit more on the database side of things first and see if there's anything else we can do in or leverage from PostgreSQL to make this work.  There are a small handful of things we could look at that'll let us know if anything is possible and whether or not we should continue on to adding more infrastructure to support this:

* Query performance tuning for schedule A and B queries
* Testing `SELECT`s across multiple cycles (read:  multiple child tables) and whether or not we need to limit how many
* Current index setup and if any others would help or hinder
* Other server settings we haven't tweaked yet but are exposed to us in RDS

I think these are all relatively lightweight things to focus on first with what we have to see if we really are pushing the boundaries of what's possible in our current setup.  If so, then I think it makes sense to start looking at something like Elasticsearch.

This work is partially related to the PR we have open to refactor the itemized schedule data that we partition (#2410) as well as issue #2394 (more directly to that one), but they're not dependent on one another.  I'm just noting that these things might be easier to accomplish with a simpler setup. :-)

**Comment by LindsayYoung** [comment]
We can explore if elastic search is an option, and improve our current setup at the same time. I don't think we need to wait on sql tuning to start exploring our options. Want to write out sql tuning as a separate issue?

**Comment by ccostino** [comment]
Okay, fair enough, I was a little worried about spreading ourselves a little thin across these but if we feel there's enough bandwidth that makes sense to me.  I'll file a separate issue for the database side.

**Comment by LindsayYoung** [comment]
I think the records we want to report are
disclosure.f_item_receipt_or_exp

**Comment by anthonygarvan** [comment]
Index: amounts, names / entities.

**Comment by anthonygarvan** [comment]
Search: by name, committee ID, date / amount filter. Occupation search. schedule type.

**Comment by LindsayYoung** [comment]
Here is the current tool
http://classic.fec.gov/finance/disclosure/norindsea.shtml

**Comment by anthonygarvan** [comment]
Setup: tony_poc_temp is a table with 20M records. 21M records loaded in elasticsearch, same schema. 

postgres:
`explain analyze select * from disclosure.tony_poc_temp where transaction_amt < 100 ORDER BY transaction_amt ASC limit 100;`
--> Takes 21.3 s

elasticsearch: 
results = Search().using(es) \
    .filter('bool', must=[Q('range', transaction_amt={'lt': 100})]) \
    .index('poc') \
    .sort('transaction_amt') \
    .extra(size=100) \
    .execute()

--> takes 0.57s

**Comment by anthonygarvan** [comment]
TL;DR --> If we have typeahead restrictions on recipient name for receipts and disbursements (so that the final query can be an equality rather than a "LIKE"), we will be able to deliver high performance queries across years, including any combination of filters, greater than / less than operators on amounts and dates, and ORDER BY clauses on any field.  

Ok, couple days and many index creations later, here's what I've found. Postgres performance really only falls into 2 categories, those where it uses indexes and returns in 5-200 ms, or those where it doesn't and returns in many minutes. 

Here's what we need to do to ensure our queries always fall into the first category:
- create multicolumn b-tree indexes on every field that gets queried. Multicolumn index consists of 2 columns: the field being queried, and the sort column (e.g., date), in that order. 
- include a LIMIT clause (100 or 1000 both worked for me)
- do not use LIKE operators in the queries. Currently we use a LIKE query for the name, that will need to be converted to a typeahead filter so that that query can become an equality once it is combined with other filters. To make the typeahead filter performant for arbitrary substring matching (eg., `LIKE '%Smith%'` and not `LIKE 'Smith%'`), it is necessary to add a trigram index to `name`. 
- ORDER BY a single column for all queries- e.g., always order by date ascending or descending.

Some query examples - 
`explain analyze select name, occupation, rpt_receipt_dt, cand_id from disclosure.f_item_receipt_or_exp where transaction_amt < 100 and rpt_receipt_dt > '1983-01-01' and cand_id='C00431569' ORDER BY transaction_amt DESC limit 100` --> 0.082 ms execution time

`explain analyze select name, occupation, rpt_receipt_dt, cand_id from disclosure.f_item_receipt_or_exp where name='TAYLOR';` --> 17ms

`explain analyze select name, occupation, rpt_receipt_dt, cand_id from disclosure.f_item_receipt_or_exp where name like '%TAYLOR%' limit 100;` --> 49 ms

`explain analyze select name, occupation, rpt_receipt_dt, cand_id from disclosure.f_item_receipt_or_exp where name like '%TAYLOR%';` --> many minutes (table scan)

`explain analyze select name, occupation, rpt_receipt_dt, cand_id from disclosure.f_item_receipt_or_exp where transaction_amt < 100 and rpt_receipt_dt > '1983-01-01' and occupation='SOFTWARE ENGINEER' limit 100;` --> 0.373ms

`explain analyze select name, occupation, rpt_receipt_dt, cand_id from disclosure.f_item_receipt_or_exp where transaction_amt < 100 and rpt_receipt_dt > '1983-01-01' and occupation='SOFTWARE ENGINEER' ORDER BY rpt_receipt_dt limit 100;` --> many minutes with single column indexes (table scan), 0.343 ms with multicolumn index defined on (occupation, rpt_receipt_dt)

I'll continue exploring this, but so far I don't see a scenario where we wouldn't get postgres to work across years with some very reasonable compromises. Also, if we wanted to include one more column to ORDER BY, for example transaction_amt, we could do that, we would just need to add another set of the multicolumn indexes on (field, transaction_amt). But that strategy starts to eat up storage quickly.

**Comment by ccostino** [comment]
@anthonygarvan thanks for all of this and the sync up this morning!  I wanted to follow up on a couple of things:

* How is a `trigram` index setup in Postgres?  I was trying to figure out which [index type](https://www.postgresql.org/docs/current/static/indexes-types.html) it corresponded to.
* We'll want to test these conditions on the partitioned tables of `ofec_sched_a_master` and `ofec_sched_b_master`.  As you had mentioned in our meeting, there shouldn't be any change in which indexes are used how, but we do want to make sure that these queries properly follow check constraints for children when doing things like searching for data across election cycles.

Let us know if you want to pair on any of this, I am happy to do so!

**Comment by anthonygarvan** [comment]
@ccostino the trigram docs are here: [](https://www.postgresql.org/docs/9.1/static/pgtrgm.html)

Here are the other indexes I created:
- CREATE INDEX tony_cand_id_idx ON disclosure.f_item_receipt_or_exp(cand_id);
- CREATE INDEX tony_rct_date_idx ON disclosure.f_item_receipt_or_exp(rpt_receipt_dt);
- CREATE INDEX tony_name_idx ON disclosure.f_item_receipt_or_exp(name);
- CREATE INDEX tony_occupation_idx ON disclosure.f_item_receipt_or_exp(occupation);-
- CREATE INDEX tony_transaction_amt_idx on disclosure.f_item_receipt_or_exp(transaction_amt);
- CREATE INDEX tony_occupation_multicolumn_idx on disclosure.f_item_receipt_or_exp(occupation, rpt_receipt_dt, transaction_amt);

Ok, I'll get start on schedules a & b :)

**Comment by anthonygarvan** [comment]
Update: I didn't realize the `name` search on the data side is a full text search, not a substring (LIKE) search. Did some testing, and ts_vector's play well with the other filters in a way that trigrams don't. eg.,
`explain analyze select * from ofec_sched_a_master where contributor_name_text @@ to_tsquery('brown') and contb_receipt_amt < 100 ORDER BY contb_receipt_dt limit 100` --> returns in 2.2 s and is a "worst case scenario" query - very common name that forces scan of the dt index.

`explain analyze select * from ofec_sched_a_master where contributor_name_text @@ to_tsquery('garvan') ORDER BY contb_receipt_dt limit 10;` --> returns in 11ms, rare name scans the full text index and filters the results afterwards

**Comment by noahmanger** [comment]
I'm closing this since the inital research spike is done and we're now focused on optimizing.
